Partial ellipsoid

srtrotter opened the pull request:
Fixes #5167

Added the ability to create a partial ellipsoid volume using new ellipsoid geometry options: innerRadii, azimuthMin, azimuthMax, elevationMin, elevationMax.

For example:
viewer.entities.add({
   position: Cesium.Cartesian3.fromDegrees(-100, 40, 5000),
   ellipsoid: {
      radii: new Cesium.Cartesian3(1500000, 1000000, 800000),
      innerRadii: new Cesium.Cartesian3(500000, 400000, 300000),
      azimuthMin: 0,
      azimuthMax: 250,
      elevationMin: 20,
      elevationMax: 70,
      material: Cesium.Color.DARKCYAN.withAlpha(0.3),
      outline: true
   }
});

[Here](https://groups.google.com/forum/embed/?place=forum/cesium-dev&showsearch=true&showpopout=true&hideforumtitle=true&fragments=true&parenturl=https%3A%2F%2Fcesiumjs.org%2Fforum%2F#!searchin/cesium-dev/ellipsoid$20volume/cesium-dev/vpWkZZ5Xvrw/5FrXxIQ1BAAJ) is the thread in the forum that this originated from.

There are a few items that might need some work, but hopefully the hard part is done.
- [x] Unit tests
- [x] Sometimes an exception is thrown when the geometry spans a pole or meridian if 2D mode is enabled. Wasn't sure what was happening. 3D works fine.
- [x] Not supported for vertexFormat of: st, tangent, bitangent. I wasn't sure what was going on there. There are TODOs in the code where this needs work.
- [x] I did not duplicate the point at the poles of the ellipsoid. I couldn't tell why this was necessary.

This work is released under U.S. DoD Distribution Statement A: Approved for public release. Distribution is unlimited.

cesium-concierge commented:
![Signed CLA is on file.](https://img.shields.io/badge/CLA-signed-brightgreen.svg)

@srtrotter, thanks for the pull request! Maintainers, we have a signed CLA from @srtrotter, so you can review this at any time.

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

srtrotter commented:
I added a bullet to CHANGES.md.

hpinkos commented:
Wow, this is great @srtrotter!  Thanks so much for the contribution! 

I was playing around with this a little bit, here are some screenshots: 

![image](https://user-images.githubusercontent.com/3451886/33088222-0e08ce98-cebb-11e7-8f1e-8afd2a1182d6.png) ![image](https://user-images.githubusercontent.com/3451886/33088230-1336ef9e-cebb-11e7-9daa-6f9ed5fe5738.png) ![image](https://user-images.githubusercontent.com/3451886/33088236-16dbf37e-cebb-11e7-9b86-1766ee262446.png)

hpinkos commented:
> Sometimes an exception is thrown when the geometry spans a pole or meridian if 2D mode is enabled

Can you paste a code snippet to reproduce this?  I wasn't able to

hpinkos commented:
>  Not supported for vertexFormat of: st, tangent, bitangent

These are used for putting textured materials on the ellipsoid.  

`st` is the texture coordinate (an (x, y) value for each vertex that maps to the x and y coordinates of an image.  These values should be between 0 and 1)

`tangent` and `bitangent` are both normalized vectors that are orthogonal to the normal vector.  For the ellipsoid, I believe we compute the `tangent` to point north an the `bitangent` is the cross product of `normal` and `tangent`

Here is some sample code to help with debugging these properties.  `DebugAppearance` colors the ellipsoid based on the texture coordinate.  `createTangentSpaceDebugPrimitive` creates a mini axis reference frame at each vertex so you can see which directions the `normal`/`tangent`/`bitangent` are pointing

```
var viewer = new Cesium.Viewer('cesiumContainer');
var scene = viewer.scene;

var radii = new Cesium.Cartesian3(200000.0, 200000.0, 200000.0);
var positionOnEllipsoid = Cesium.Cartesian3.fromDegrees(-100.0, 40.0);
var modelMatrix = Cesium.Matrix4.multiplyByTranslation(
    Cesium.Transforms.eastNorthUpToFixedFrame(positionOnEllipsoid),
    new Cesium.Cartesian3(0.0, 0.0, radii.z), new Cesium.Matrix4()
);

var ellipsoidGeometry = new Cesium.EllipsoidGeometry({
    vertexFormat : Cesium.VertexFormat.ALL,
    radii : radii
});
var ellipsoidInstance = new Cesium.GeometryInstance({
    geometry : ellipsoidGeometry,
    modelMatrix : modelMatrix,
});
scene.primitives.add(new Cesium.Primitive({
    geometryInstances : ellipsoidInstance,
    appearance : new Cesium.DebugAppearance({
        attributeName: 'st',
        translucent : false,
        closed : true
    })
}));
scene.primitives.add(Cesium.createTangentSpaceDebugPrimitive({
 geometry : ellipsoidGeometry,
 length : 10000.0,
 modelMatrix : modelMatrix
}));
```

![image](https://user-images.githubusercontent.com/3451886/33089009-60c648fc-cebd-11e7-960e-95d7ac672d07.png)

hpinkos commented:
You have some indicies winding order issues:

![image](https://user-images.githubusercontent.com/3451886/33089577-028f1370-cebf-11e7-942e-7c5b85825716.png)

This is happening because some of the triangles are facing outward and some are facing in.  The triangles should all have a counter-clockwise winding order

In this example, the indicies for the two triangles would be (1, 2, 3) and (1, 3, 4)

![image](https://user-images.githubusercontent.com/3451886/33089768-a38d0d54-cebf-11e7-9f68-1192393b2435.png)


Here's the code to reproduce:

```
var viewer = new Cesium.Viewer('cesiumContainer');
var scene = viewer.scene;

var radii = new Cesium.Cartesian3(200000.0, 200000.0, 300000.0);
var positionOnEllipsoid = Cesium.Cartesian3.fromDegrees(-100.0, 40.0);
var modelMatrix = Cesium.Matrix4.multiplyByTranslation(
    Cesium.Transforms.eastNorthUpToFixedFrame(positionOnEllipsoid),
    new Cesium.Cartesian3(0.0, 0.0, radii.z), new Cesium.Matrix4()
);
var ellipsoidGeometry = new Cesium.EllipsoidGeometry({
    vertexFormat : Cesium.PerInstanceColorAppearance.VERTEX_FORMAT,
    radii: new Cesium.Cartesian3(1500000, 1000000, 800000),
    innerRadii: new Cesium.Cartesian3(500000, 400000, 300000),
    azimuthMin: 0,
    azimuthMax: 250,
    elevationMin: 20,
    elevationMax: 70,
});
var ellipsoidInstance = new Cesium.GeometryInstance({
    geometry : ellipsoidGeometry,
    modelMatrix : modelMatrix,
    attributes : {
        color : Cesium.ColorGeometryInstanceAttribute.fromColor(Cesium.Color.CYAN)
    }
});
scene.primitives.add(new Cesium.Primitive({
    geometryInstances : ellipsoidInstance,
    appearance : new Cesium.PerInstanceColorAppearance({
        translucent : false,
        closed: true
    })
}));
```

hpinkos commented:
The normals for the inner ellipsoid are pointing the wrong way:

![image](https://user-images.githubusercontent.com/3451886/33089920-284dff62-cec0-11e7-9122-44897fbc4fe8.png)

I would expect them to be pointing in towards the center

hpinkos commented:
Thanks again for the pull request @srtrotter!  This is a pretty common feature request and our community will be thrilled once it is finally merged in =)

This are my initial comments for now.  I'll take a closer look at the code after those few things are addressed, but this is a great start!  Let me know if you have any questions or need clarification on anything.

srtrotter commented:
Thanks for the info, @hpinkos ! I figured I had some issues like these that weren't exposed for my use cases. It will be next week at the earliest before I can look into any of these.

hpinkos commented:
@srtrotter no problem!  Hope you have a good Thanksgiving

srtrotter commented:
To trigger the 2D exception:

`viewer.entities.add({
   position: Cesium.Cartesian3.fromDegrees(-100, 90, 5000),
   ellipsoid: {
      radii: new Cesium.Cartesian3(1500000, 1000000, 800000),
      innerRadii: new Cesium.Cartesian3(500000, 400000, 300000),
      azimuthMin: 0,
      azimuthMax: 250,
      elevationMin: 20,
      elevationMax: 90,
      material: Cesium.Color.DARKCYAN.withAlpha(0.3),
      outline: true
   }
});`

srtrotter commented:
@hpinkos, could you please provide a code example to reproduce the 2D texture blend image, regarding duplicating the point at the poles?

pjcozzi commented on Source/DataSources/CzmlDataSource.js:
```diff
@@ -1534,6 +1534,11 @@ define([
         processPacketData(Boolean, ellipsoid, 'show', ellipsoidData.show, interval, sourceUri, entityCollection, query);
         processPacketData(Cartesian3, ellipsoid, 'radii', ellipsoidData.radii, interval, sourceUri, entityCollection, query);
+        processPacketData(Cartesian3, ellipsoid, 'innerRadii', ellipsoidData.innerRadii, interval, sourceUri, entityCollection);
+        processPacketData(Number, ellipsoid, 'azimuthMin', ellipsoidData.azimuthMin, interval, sourceUri, entityCollection);
```
Throughout, please write out `Minimum` and `Maximum`.

We avoid abbreviations in the public API: https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#naming

pjcozzi replied:
Also, to be consistent the rest of Cesium, make "minimum" and "maximum" the prefix instead of the suffix, e.g., `minimumAzimuth`.

srtrotter commented:
@hpinkos, nevermind. I see how to draw the image.

pjcozzi commented on CHANGES.md:
```diff
@@ -6,6 +6,7 @@ Change Log
 * Added ability to support touch event in Imagery Layers Split demo application. [#5948](https://github.com/AnalyticalGraphicsInc/cesium/pull/5948)
 * Fixed `Invalid asm.js: Invalid member of stdlib` console error by recompiling crunch.js with latest emscripten toolchain. [#5847](https://github.com/AnalyticalGraphicsInc/cesium/issues/5847)
 * Added CZML support for `polyline.depthFailMaterial`, `label.scaleByDistance`, `distanceDisplayCondition`, and `disableDepthTestDistance`. [#5986](https://github.com/AnalyticalGraphicsInc/cesium/pull/5986)
+* Added ability to create partial ellipsoid volumes using new geometry options: inner radii, min/max azimuth, min/max elevation.
```
Can this explicitly mention the new Cesium classes and CZML properties?  And link to the Sandcastle example if you are able to add one?

pjcozzi commented on Source/Core/EllipsoidGeometry.js:
```diff
@@ -50,6 +50,11 @@ define([
      * @param {Object} [options] Object with the following properties:
      * @param {Cartesian3} [options.radii=Cartesian3(1.0, 1.0, 1.0)] The radii of the ellipsoid in the x, y, and z directions.
+     * @param {Cartesian3} [options.innerRadii=options.radii] The inner radii of the ellipsoid in the x, y, and z directions.
+     * @param {Number} [options.azimuthMin=0.0] The minimum azimuth in degrees (0 is north, +CW).
```
Please use radians throughout to be consistent with the rest of Cesium.

srtrotter commented:
@pjcozzi I am working on those changes and yes, I should be able to put a Sandcastle example together.

srtrotter commented:
@hpinkos I made some updates based on the feedback from @pjcozzi. I also fixed the indices winding order issue. Can you please verify it is working as you would expect? Still working on the other issues.
Note that the inputs are in radians now, so you'll have to update your code that creates an ellipsoid.

hpinkos commented:
Just took a look @srtrotter, and the winding order looks good to me!

srtrotter commented:
@hpinkos, I believe I have fixed the issue with the normals and the texture seam. Please take a look and let me know if everything is working like you would expect.

I think the only outstanding issue is the exception thrown by CombineGeometry when 2D mode is enabled for an ellipsoid with a maximumAzimuth of 90 degrees that crosses a pole. Is that something that someone on your team can look into?

I will also work on adding a Sandcastle example.

mramato commented:
Looks like this PR is in good hands, I just wanted to stop by and say this is awesome and thank you @srtrotter!  A ton of people will be really excited about this (including me!).

@shunter this introduces some new CZML properties that should be a breeze to add, if you don't beat me too it I'll try and be a good Samaritan and open a PR for it after this is merged.

srtrotter commented:
You're welcome @mramato! Glad to be able to contribute. I did add CZML support for the new ellipsoid properties. Not sure if you're referring to something else, though.

srtrotter commented:
@hpinkos I fixed another little issue with the normals. I think there is some additional work that could be done, to rotate the normals on the surfaces that connect the outer and inner ellipsoids. That gets complicated quickly and I can't really afford to spend time on it right now. Perhaps that could be marked as a future improvement.

mramato commented:
@srtrotter I did notice that (and was happy to see it).  This is difference, we have JSON schema definition for CZML that generates the CZML specification/documentation as well as C# and Java CZML writers.  (All managed in a separate repository: https://github.com/AnalyticalGraphicsInc/czml-writer)

srtrotter commented:
@pjcozzi I updated the change entry. I also updated the Spheres and Ellipsoids Sandcastle example. Let me know if you'd like anything changed/added/removed.

srtrotter commented:
@hpinkos I'm having trouble figuring out why the CI tests are failing. Are you able to tell why?

hpinkos commented:
@srtrotter our CI tests have been failing intermittently.  As long as they pass when you run them locally, don't worry about it.

I'll try to take another look at this PR next week, sorry it's been a bit busy around here

srtrotter commented:
@hpinkos just wanted to check in and see if you had a chance to look at this any more.

hpinkos commented:
Sorry for not getting back to this yet @srtrotter.  Things have been crazy here trying to wrap things up before people leave for Christmas vacation.  I should have some time in the next two weeks.  Thanks for the reminder!

srtrotter commented:
No problem! I understand how things can get crazy at the end of the year. Just wanted to make sure you weren't waiting on me for something.

srtrotter commented:
@hpinkos Something to consider when you review this next: I think the azimuth/elevation variable names I chose might need to be changed. They are really just the yaw/pitch angles relative to the entity's coordinate system. I wanted to get some input before I made any changes.

hpinkos commented:
@srtrotter hmm, I'm not sure which variable names would be most familiar to people in this context.  We use heading/pitch/roll elsewhere in the code base.   @pjcozzi do you have any insight here?

pjcozzi commented:
> I think the azimuth/elevation variable names I chose might need to be changed. They are really just the yaw/pitch angles relative to the entity's coordinate system. I wanted to get some input before I made any changes.

Can we make it consistent with Cesium's [Spherical](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Spherical.js) type?

>     * @param {Number} [clock=0.0] The angular coordinate lying in the xy-plane measured from the positive x-axis and toward the positive y-axis.
>     * @param {Number} [cone=0.0] The angular coordinate measured from the positive z-axis and toward the negative z-axis.

pjcozzi commented:
@srtrotter after you make the next round of changes, can you please run eslint and fix the issues?
```
  412:21  error  'innerOffset' used outside of binding context  block-scoped-var
  416:40  error  'innerOffset' used outside of binding context  block-scoped-var
  419:40  error  'innerOffset' used outside of binding context  block-scoped-var
  420:40  error  'innerOffset' used outside of binding context  block-scoped-var
  426:21  error  'outerOffset' used outside of binding context  block-scoped-var
  427:17  error  'innerOffset' used outside of binding context  block-scoped-var
  427:17  error  'innerOffset' used outside of binding context  block-scoped-var
  429:40  error  'outerOffset' used outside of binding context  block-scoped-var
  430:40  error  'outerOffset' used outside of binding context  block-scoped-var
  431:40  error  'innerOffset' used outside of binding context  block-scoped-var
  431:40  error  'innerOffset' used outside of binding context  block-scoped-var
  433:40  error  'outerOffset' used outside of binding context  block-scoped-var
  434:40  error  'innerOffset' used outside of binding context  block-scoped-var
  434:40  error  'innerOffset' used outside of binding context  block-scoped-var
  435:40  error  'innerOffset' used outside of binding context  block-scoped-var
  435:40  error  'innerOffset' used outside of binding context  block-scoped-var
  442:17  error  'outerOffset' is already defined               no-redeclare
  443:17  error  'innerOffset' is already defined               no-redeclare
  445:17  error  'innerOffset' used outside of binding context  block-scoped-var
  446:17  error  'outerOffset' used outside of binding context  block-scoped-var
  447:36  error  'innerOffset' used outside of binding context  block-scoped-var
  448:36  error  'outerOffset' used outside of binding context  block-scoped-var
  449:36  error  'outerOffset' used outside of binding context  block-scoped-var
  451:36  error  'innerOffset' used outside of binding context  block-scoped-var
  452:36  error  'innerOffset' used outside of binding context  block-scoped-var
  453:36  error  'outerOffset' used outside of binding context  block-scoped-var
  457:17  error  'innerOffset' used outside of binding context  block-scoped-var
  458:17  error  'outerOffset' used outside of binding context  block-scoped-var
  459:36  error  'outerOffset' used outside of binding context  block-scoped-var
  460:36  error  'innerOffset' used outside of binding context  block-scoped-var
  461:36  error  'outerOffset' used outside of binding context  block-scoped-var
  463:36  error  'outerOffset' used outside of binding context  block-scoped-var
  464:36  error  'innerOffset' used outside of binding context  block-scoped-var
  465:36  error  'innerOffset' used outside of binding context  block-scoped-var

/Users/pcozzi/source/cesium/Apps/Sandcastle/gallery/Spheres and Ellipsoids.html
  163:60  error  Parsing error: Unexpected token ,

✖ 35 problems (35 errors, 0 warnings)
```

For more on running eslint, see the [Build Guide](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/BuildGuide#build-scripts).  This is also ran in CI when you push to your branch in this pull request.

pjcozzi commented on Apps/Sandcastle/gallery/Spheres and Ellipsoids.html:
```diff
@@ -62,6 +62,178 @@
+    name : 'Airplane with sensor',
+    position : airplanePosition,
+    model : {
+        "../../SampleData/models/CesiumAir/Cesium_Air.gltf",
```
This has a syntax error on load.
```
Uncaught SyntaxError: Unexpected token , (on line 134)
```

pjcozzi replied:
There's also a few more like this in this file.

pjcozzi replied:
Probably just remove the model entities from the example.

srtrotter replied:
@pjcozzi I'd really like to leave the model entities in the example, unless you feel strongly that they shouldn't be included. I think it would be very helpful to provide examples for how to combine a model with a partial ellipsoid.

pjcozzi replied:
I really do not think they add value - keep the example focused.

pjcozzi commented on Apps/Sandcastle/gallery/Spheres and Ellipsoids.html:
```diff
@@ -4,7 +4,7 @@
     <meta http-equiv="X-UA-Compatible" content="IE=edge">
     <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
-    <meta name="description" content="Draw spheres and ellipsoids.">
+    <meta name="description" content="Draw spheres, ellipsoids, and sensor volumes.">
```
Very cool example.

Please update the screenshot in the same directory.

Also, let's not call these "sensor volumes" since that means something quiet different to the AGI community using Cesium.  Perhaps partial ellipsoids or ellipsoid slices?

pjcozzi replied:
Also, if you want to go wild in this example, it could be cool to have an example or two that animates!

srtrotter replied:
@pjcozzi What if I change it to "sensor coverage"? A use case for this ability is to draw the coverage area of a radar, so I wanted to mention that somehow.

pjcozzi replied:
Please keep this as generic capability - the Cesium team would not endorse this as sensors - as we do a lot of work in this area with fully featured sensor visibility models and can not confuse our user community.

pjcozzi commented on CHANGES.md:
```diff
@@ -13,6 +13,7 @@ Change Log
 * Fixed globe materials when `Globe.enableLighting` was `false`. [#6042](https://github.com/AnalyticalGraphicsInc/cesium/issues/6042)
 * Fixed shader compilation failure on pick when globe materials were enabled. [#6039](https://github.com/AnalyticalGraphicsInc/cesium/issues/6039)
 * Fixed crash when `invertClassification` was enabled, the invert color had an alpha less than `1.0`, and the window was resized. [#6046](https://github.com/AnalyticalGraphicsInc/cesium/issues/6046)
+* Added ability to create partial ellipsoid volumes using both the Entity API and CZML. New ellipsoid geometry properties: innerRadii, minimumAzimuth, maximumAzimuth, minimumElevation, maximumElevation. See the updated [Sandcastle example](https://cesiumjs.org/Cesium/Apps/Sandcastle/?src=Spheres%20and%20Ellipsoids.html&label=Geometries). [#5995](https://github.com/AnalyticalGraphicsInc/cesium/pul ...
```
Perhaps move this to a new 1.42 section (release date: 2018-02-01); I do not know that we will have enough time to review this for 1.41 given the holidays.

pjcozzi commented on Source/Core/EllipsoidGeometry.js:
```diff
@@ -50,6 +50,11 @@ define([
      *
      * @param {Object} [options] Object with the following properties:
      * @param {Cartesian3} [options.radii=Cartesian3(1.0, 1.0, 1.0)] The radii of the ellipsoid in the x, y, and z directions.
+     * @param {Cartesian3} [options.innerRadii=options.radii] The inner radii of the ellipsoid in the x, y, and z directions.
```
Can you also mention updates to this and `EllipsoidOutlineGeometry` in CHANGES.md?

pjcozzi commented on Source/Core/EllipsoidGeometry.js:
```diff
@@ -50,6 +50,11 @@ define([
      * @param {Object} [options] Object with the following properties:
      * @param {Cartesian3} [options.radii=Cartesian3(1.0, 1.0, 1.0)] The radii of the ellipsoid in the x, y, and z directions.
+     * @param {Cartesian3} [options.innerRadii=options.radii] The inner radii of the ellipsoid in the x, y, and z directions.
+     * @param {Number} [options.minimumAzimuth=0.0] The minimum azimuth in radians.
```
For each new angle parameter here and in `EllipsoidOutlineGeometry`, can the reference doc please include the expected range?  And then throw a `DeveloperError` when the input is out of range?

srtrotter replied:
@pjcozzi I'd rather not enforce a specific range, simply because the code can handle any range. It will simply "wrap" the angles if necessary. This also allows you to create some unique shapes if you get creative with the inputs. I'd like to preserve that ability. Let me know if you disagree.

pjcozzi replied:
OK, just document it please.

pjcozzi commented:
@srtrotter could you add unit tests?  It should be easy to get started by adding to the existing tests for ellipsoid/outline geometries and entities.  For more on running and writing tests in Cesium, see the [Testing Guide](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide#testing-guide).

pjcozzi commented:
Teaser screenshot for those interested in @srtrotter's nice work:

![image](https://user-images.githubusercontent.com/782098/34280606-ef05f15e-e686-11e7-8423-6cbdb557c295.png)

srtrotter commented:
Thanks for the feedback @pjcozzi! It will probably be next week before I can get to this, but I should be able to make all of the changes.

pjcozzi commented:
Great, no rush, thanks for making the updates!

srtrotter commented:
@pjcozzi, @hpinkos I have made the following changes. Please note that I am doing this development on a machine without internet access. Adding new software is difficult, time consuming, and requires lots of hoop jumping. So unfortunately, I am unable to do things like run eslint and the unit tests right now. Getting changes from that machine to over here is also difficult and error prone. I took a stab at addressing all of your findings, but I'll apologize in advance for any errors that remain.

I think there are still a few things that will need to be addressed. I'm afraid I have spent as much time as I can afford on this for the time being, except for trivial findings. Is there any way someone on your team could tie up any loose ends?

* Renamed azimuth/elevation to clock/cone and changed the convention to be consistent with the Spherical type.
* I believe I fixed the eslint issues. Unfortunately, I can't run eslint as explained above.
* Fixed the errors in the sandcastle example and made a few modifications. I'm afraid I'll need one of you to update the screenshot for me.
* Added unit tests. Since I can't run the unit tests, there might be errors in there (sorry!). The new test coverage is not 100%, but I tried to cover the basics, given the restrictions I'm under right now.
* Note that I removed a section from EllipsoidGeometry (st VertexFormat adjustment near poles). It was causing me some issues and I was unable to figure out its purpose. I'm sure it needs to be in there, but I didn't understand what it needed to do and didn't have any more time to investigate.

I'm sorry to drop an unfinished product on you. Let me know what you are able to do, and I'll do my best to work with you.

Thanks!

hpinkos commented:
No problem @srtrotter!  We really appreciate all the work you've done towards this and we have no problem finishing it up.  Unless someone else gets to it first, I'll plan on finishing this up next time our team has a code sprint

srtrotter commented:
@hpinkos Ok great! Any idea when that might be? When you do get around to it, let me know if you have any questions about anything that I did. I'll try to help as much as I can.

hpinkos commented:
@srtrotter we don't have anything on the calendar yet, but in the next few months.  Thanks again!

srtrotter commented:
@hpinkos just wanted to drop in and see if this was going to be finished up in the near future. Thanks!

hpinkos commented:
Hi @srtrotter!  Sorry, we've been really busy around here getting ready for a big milestone on May 1st.  I'm hoping to have a little time to work on this after then, but I'm not sure.

srtrotter commented:
@hpinkos, @pjcozzi, not trying to be a nuisance, but just wanted to drop in and see if there were any plans to try to merge this in anytime soon. Would love for it to be in the main line!

hpinkos commented:
Hi @srtrotter.  Sorry, but we're a pretty small team and we just haven't been able to make this a priority yet.  I would love to get this merged in master too.

srtrotter commented:
@hpinkos no worries! I understand. Just wanted to check.

cesium-concierge commented:
Thanks again for the pull request!

Looks like this pull request hasn't been updated in 30 days since I last commented.

To keep things tidy should this be closed? Perhaps keep the branch and submit an issue?
---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

hpinkos commented:
Sorry @mcrummey, we haven't been able to make this a priority yet.  If you have time to take over and get it over the finish line, we would be happy to review the pull request!

nahgrin commented:
I am willing to pick up an finish this task.

What is all that needs to be done and is there procedures I should follow when addressing them?

Thank you in advance!

hpinkos commented:
@nahgrin thanks for volunteering to finish this up!  There have been some changes made to `EllipsoidGeometry` since this PR was first opened, so the first step would be merging master into this branch.

After that, I think the main thing that was holding this up from being merged was a lack of unit tests.  I didn't have a chance to run the tests that @srtrotter added and I'm not sure whether or not they're passing.

Other than that, I was planning to double check that all of the vertex attributes were being computed correctly in all cases (st, normal, tangent, binormal).
And there is a crash that's happening in 2D that needs to be addressed https://github.com/AnalyticalGraphicsInc/cesium/pull/5995#issuecomment-347221274

srtrotter commented:
@nahgrin that is great! If you have any questions about anything that I did, don't hesitate to ask. Thanks for picking this up!

nahgrin commented:
@hpinkos @srtrotter 

I'm in the process of merging master into the Partial Ellipsoid branch. I'm having issues seeing any partial ellipsoids though, there were a few merge issues that I believe were resolved properly.

I've verified that all the new options are getting to the create geometry in EllipsoidGeometry.js, however; I'm unsure if it's actually drawing the geometry correctly. I'm just getting a regular ellipsoid when I try to add partial ellipsoids in a sand castle.

nahgrin commented:
@srtrotter @hpinkos 

I've been in the process of merging the master branch into @srtrotter 's branch.

I believe I've resolved most of the merge conflicts. However, I'm having issues with the shape fills that I suspect are coming from the EllipsoidGeometry.js. It seems as though the indices are off and that code has changed quite a bit since this branch was created.

Any ideas?

Below are pictures of what I'm seeing:
![example1](https://user-images.githubusercontent.com/26147898/46557324-17dd2100-c8af-11e8-804a-b48812329398.PNG)
![example2](https://user-images.githubusercontent.com/26147898/46557328-19a6e480-c8af-11e8-82e4-461dba8a257e.PNG)

srtrotter commented:
@nahgrin Yeah it looks like the indices are wrong when you're creating the triangles. Looks like you might only be off by 1. The only thing I know to do is to carefully compare the old vs the new. I can probably take a look later this week. Do you have this work in a branch somewhere?

nahgrin commented:
@srtrotter 

I've managed to push up code to a fork of your repo: https://github.com/nahgrin/cesium/tree/upstream/partial-ellipsoid

I'm having issues getting a specific test to pass. It's the 'computes attributes for a unit sphere' in EllipsoidGeometry, it seems that the dot product of the z unit and the tangent is not always greater than or equal to zero.

I was hoping you had some insight on the math portion of that :)

srtrotter commented:
@nahgrin sorry for the delay. It's possible/likely those unit tests were not passing to begin with, and may not be anything that you changed in the merge. That might also be the reason I was having trouble with the 2D map. Is it always the poles of the ellipsoid that are the problem?

nahgrin commented:
@srtrotter In manual steps when putting the ellipsoid on the poles in  a sandcastle it would still crash. I believe the test that is failing is not dealing with the poles in any way. However, I have a hunch that they might be related somehow.

srtrotter commented:
@nahgrin Sorry, I meant what indices is it complaining about regarding the dot product? Is it the indices of the poles of the ellipsoid? I wasn't referring to the poles of the earth.

nahgrin commented:
@srtrotter Sorry about the delay, I got pulled away from this work for the time being. But I'm going to be working on this a lot in the coming weeks. I'll identify the indices and post them back here.

Thank you again.

nahgrin commented:
@srtrotter @hpinkos 

I believe I've solved the 2d crashing, I had to update the barycentricCoordinates to prevent the points from dividing by infinity. You can see those changesets here: https://github.com/nahgrin/cesium/commits/upstream/partial-ellipsoid

Just wanted to make sure that's an okay fix?

I also updated some tests to full coverage and fixed es-lint issues. I'll continue to update tests and will try to fix that one that is failing. When I gather more information on why it's failing I'll post back here.

srtrotter commented:
@nahgrin That's great! Sounds like you're making good progress. Your question about the barycentricCoordinates really needs to be answered by @hpinkos or someone else on the dev team. I can't speak to that one way or another.

Do you tests cover the full range of options for the min/max clock/cone angles? You'll want to make sure you test all different combinations, if you haven't already.

Thanks again for picking this up. I'm really excited about this getting merged in!

nahgrin commented:
@srtrotter I'm going to be working on finishing up the rest of the code to get it in order to submit back into master.

Would you prefer that I submit into your PR or would you rather I submit from my upstream PR?

srtrotter commented:
@nahgrin It might be best if you submit into this PR in order to better preserve the history and conversation here. But I'm fine if you submit from yours, if you want to. If you submit into mine, make sure you add yourself to the contributors.

nahgrin commented:
@srtrotter I'd like to submit against yours and I'll add myself to the contributors. I'll be working a lot on getting this good to go this coming week.

Thank you again. :)

srtrotter commented:
@nahgrin Sounds good!

hpinkos commented:
@nahgrin yes, we'll take care of updating the czml writer after these new features get merged into master.  Looking forward to your PR!  Thanks for taking this over =)

nahgrin commented:
Hey all (@srtrotter , @pjcozzi , @hpinkos , @shunter ),

I'm running into some management issues on the contributors guide update. I'm incredibly sorry for the wait...

hpinkos commented:
No worries @nahgrin, thanks for the update!  Thanks again for working on this, and let us know if we can help.

srtrotter commented:
@nahgrin Just wanted to see how things were going. Any progress on the issues you were facing with the contributors guide?

nahgrin commented:
@srtrotter Unfortunately no, still waiting on a response for it. 😞 All my code except for the contributors guide should be on my upstream branch.

cesium-concierge commented:
Thanks again for your contribution @srtrotter! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @srtrotter! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

srtrotter commented:
@nahgrin Any luck with resolving the red tape?

cesium-concierge commented:
Thanks again for your contribution @srtrotter! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @srtrotter! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

nahgrin commented:
@srtrotter @hpinkos   I'm back! This time with a way forward. I'm going to re-merge master into my branch.

To save time, I'll submit from my branch when the code is good to go.

Thank you all for your patience!

srtrotter commented:
@nahgrin Excellent! Looking forward to seeing it merged in!

nahgrin commented:
@hpinkos @srtrotter ,

Since my code is an upstream of @srtrotter upstream, what would be the process I would need to do get my upstream to push to master?

Do I need to open a new PR? Or would it be easier just to push it to @srtrotter upstream? Whatever is faster I'm okay with :)

srtrotter commented:
@nahgrin Ok you should have permissions now! Let me know if you need anything else.

nahgrin commented:
@srtrotter It looks as though I'm still getting permission denied when trying to push.

We're running the command: git push origin partial-ellipsoid
from our repo and getting a 403.

srtrotter commented:
@nahgrin I sent you an invite to be a collaborator: https://github.com/srtrotter/cesium/invitations

It's saying that it's still waiting for your response. Try accepting it and then try to push again. Let me know if that still doesn't work.

nahgrin commented:
@srtrotter I pushed the code into your branch. It's now showing up as part of this PR.

@hpinkos @pjcozzi @shunter Would you be able to look things over to make sure we're good?

srtrotter commented:
@hpinkos Do you need anything further me me or @nahgrin? Would really love to get this merged!

hpinkos commented:
Thanks for finishing this up @nahgrin !

@srtrotter @nahgrin I'm not sure now soon I'll be able to review this.  We have a project with a pretty tight deadline right now, but I'll try to carve out some time.

@nahgrin if you haven't already, please send in a [Contributor License Agreement](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) since you added commits to this PR.  Thanks!

nahgrin commented:
@hpinkos 

I've been informed that the CLA has been sent to you guys on 5/30/19. I believe it had been sent to @pjcozzi .

pjcozzi commented:
Ah, yes, we have a CLA just for this specific pull request so it is not in our automated systems.  Thanks and sorry!

nahgrin commented:
@pjcozzi No problems at all. Let me know if there's anything additional @srtrotter and I could do to help things get merged :)

nahgrin commented:
@pjcozzi @hpinkos 

Any updates? How's the deadline stuff going?

cesium-concierge commented:
Thanks again for your contribution @srtrotter! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

nahgrin commented:
@hpinkos 
@pjcozzi 

Sorry to keep pinging you guys, any status on getting this merged in?

hpinkos commented:
Sorry we haven't had a chance to review this one yet @nahgrin.  I'm not sure exactly when I'll have a chance, but it shouldn't be much longer

cesium-concierge commented:
Thanks again for your contribution @srtrotter! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

hpinkos commented:
Thank you so much for the fantastic work on this @srtrotter and @nahgrin!  I'm sorry it took so long for me to take a final look at this.  I just tested everything and it looks great.  I just made some minor changes to the Sandcastle example. 

I'm going to merge this as soon as CI passes

nahgrin commented:
@hpinkos thank you so much for all your help!

I'm so excited! :)

And thank you @srtrotter for starting this and helping getting it to finish.

hpinkos commented:
🎉 🎊 💯

hpinkos commented:
@shunter FYI new CZML properties for `ellipsoid`: `innerRadii`, `minimumClock`, `maximumClock`, `minimumCone`, and `maximumCone`

srtrotter commented:
Great! Thanks @hpinkos for getting this merged in! And thanks @nahgrin for your work in getting this through to the finish line! Very excited that this is complete.

vince-hunt commented:
The implementation of this appears a bit buggy.  Clipping against the terrain is not consistent at different zoom levels.

https://sandcastle.cesium.com/?src=Partial%20Ellipsoids.html

Zoomed out:
![image](https://user-images.githubusercontent.com/46467654/78322127-2219d180-75b1-11ea-84b8-336886d09410.png)

Zoomed in:
![image](https://user-images.githubusercontent.com/46467654/78322168-470e4480-75b1-11ea-90eb-534169a23f38.png)

This clipping issue affects all the partial ellipsoids.  It is particularly noticeable in the wedge shape shown above, but it can be seen on other ellipsoids where the ellipsoid meets the earth.

srtrotter commented:
@vince-hunt I don't think it has anything to do with the ellipsoid being partial. This problem exists (and has always existed) for all ellipsoids. I see it on full spheres as well. Are you seeing something that's different for a partial ellipsoid?

serdaryesilmurat commented:
A little late for the conversation 😅 , but is it possible to draw the sector (3rd geometry) shown in the picture below?
The "Wedge" sample in https://sandcastle.cesium.com/?src=Partial%20Ellipsoids.html (the darkcyan partial ellipsoid) does not cover what we need actually.

> Wow, this is great @srtrotter! Thanks so much for the contribution!
> 
> I was playing around with this a little bit, here are some screenshots:
> 
> ![image](https://user-images.githubusercontent.com/3451886/33088222-0e08ce98-cebb-11e7-8f1e-8afd2a1182d6.png) ![image](https://user-images.githubusercontent.com/3451886/33088230-1336ef9e-cebb-11e7-9daa-6f9ed5fe5738.png) ![image](https://user-images.githubusercontent.com/3451886/33088236-16dbf37e-cebb-11e7-9b86-1766ee262446.png)

srtrotter commented:
```JavaScript
viewer.entities.add({
  name: "C",
  position: Cesium.Cartesian3.fromDegrees(-102.0, 35.0, 150000.0),
  ellipsoid: {
    radii: new Cesium.Cartesian3(500000.0, 500000.0, 800000.0),
    innerRadii: new Cesium.Cartesian3(400000.0, 400000.0, 400000.0),
    minimumClock: Cesium.Math.toRadians(45.0),
    maximumClock: Cesium.Math.toRadians(315.0),
    minimumCone: Cesium.Math.toRadians(80.0),
    maximumCone: Cesium.Math.toRadians(100.0),
    material: Cesium.Color.DARKCYAN.withAlpha(0.3),
    outline: true,
  },
});
```
@serdaryesilmurat that should get you something similar. You may have to create multiple ellipsoids and place them next to each other to get the exact shape you're looking for. Not sure what @hpinkos used to create that original shape.

serdaryesilmurat commented:
@srtrotter Yes, that's what I am looking for 😃  
Really appriciated for the quick response 😉
